### PR TITLE
Add cog for turning math expressions into LaTeX 

### PR DESCRIPTION
**Pull request by martmists**
Note that this does NOT require eval, and does NOT evaluate the expression, and is therefore 100% safe.

### CODE REVIEW

File: bot/__main__.py [hunk 1]
```diff
@@ -34,5 +34,6 @@
 bot.load_extension("bot.cogs.fun")
 bot.load_extension("bot.cogs.eval")
 bot.load_extension("bot.cogs.verification")
+bot.load_extension("bot.cogs.math")
```
Comment (AnonGuy) [inline bot/__main__.py]: Could this be on the previous line? Not a must-have, but these lines would certainly look nicer with ascending lengths :D

File: bot/cogs/math.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+        else:
+            ltx = latex(parsed)
+            urlsafe = urlescape(ltx)
+
+            async with ClientSession() as session:
+                async with session.get(LATEX_URL + urlsafe) as resp:
+                    bytes_img = await resp.read()
+
+            await ctx.send(file=File(fp=BytesIO(bytes_img), filename="latex.png"))
```
Comment (AnonGuy) [inline bot/cogs/math.py]: Might we have this `File(fp=BytesIO(bytes_img)` on a previous line. Looks a bit too condensed to me.

**Comment by AnonGuy** [comment]
The current font that the LaTeX is rendered with is Helvetica. Any way to change the request to use the Latin Modern font? Seems to be the one used in most LaTeX rendered expressions

File: bot/cogs/math.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+
+from io import BytesIO
+from urllib.parse import urlescape
```
Comment (gdude2002) [inline bot/cogs/math.py]: This doesn't exist

**Comment by hargoniX** [comment]
There is https://en.wikibooks.org/wiki/LaTeX/Fonts

**Comment by AnonGuy** [comment]
It looks like the font is passed directly in the URL as a query string: `fn_phv`. In my opinion, Latin Modern (`fn_lm`) looks much nicer.

File: bot/cogs/math.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+from aiohttp import ClientSession
+
+from discord import File
+from discord.ext.commands import command
+
+from sympy import latex
+from sympy.parsing.sympy_parser import parse_expr
+
+
+LATEX_URL = "https://latex.codecogs.com/png.download?%5Cdpi%7B150%7D%20%5Cbg_white%20%5Cfn_phv%20"
```
Comment (AnonGuy) [inline bot/cogs/math.py]: The return PNG from this base URL is much too small. Change it to something like `https://latex.codecogs.com/gif.latex?%5Cdpi%7B300%7D%20%5Cbg_white%20%5Clarge%20ABC`
e.g add `300` and `large` to the query string to make it readable.

**Comment by martmists** [comment]
- [ ] cog order by lenght
- [ ] larger image and Latin Modern
- [ ] fix urlescape
- [ ] sympy in requirements.txt
- [ ] Shorten file send line

I will fix these issues tomorrow

File: bot/cogs/math.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+from aiohttp import ClientSession
+
+from discord import File
+from discord.ext.commands import command
+
+from sympy import latex
+from sympy.parsing.sympy_parser import parse_expr
+
+
+LATEX_URL = "https://latex.codecogs.com/png.download?%5Cdpi%7B300%7D%20%5Cbg_white%20%5Chuge%20"
```
Comment (AnonGuy) [inline bot/cogs/math.py]: Consider removing `bg_white` from the query string completely, to return an expression with a transparent background. You might also want to decrease that text size a little. `150` seems to work well in discord.

Comment (martmists) [reply bot/cogs/math.py]: bg_white works better for users on dark or custom themes, I will change the text size though

File: bot/cogs/math.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+from discord.ext.commands import command
+
+from sympy import latex
+from sympy.parsing.sympy_parser import parse_expr
+
+
+LATEX_URL = "https://latex.codecogs.com/png.download?%5Cdpi%7B150%7D%20%5Cbg_white%20%5Chuge%20"
+
+
+class MathCog:
```
Comment (gdude2002) [inline bot/cogs/math.py]: Should be named "Math"

Should have an `__init__()` that takes a `bot` object and stores it under `self

File: bot/cogs/math.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+            ltx = latex(parsed)
+            urlsafe = quote(ltx)
+
+            async with ClientSession() as session:
+                async with session.get(LATEX_URL + urlsafe) as resp:
+                    bytes_img = await resp.read()
+
+            file = File(fp=BytesIO(bytes_img), filename="latex.png")
+
+            await ctx.send(file=file)
```
Comment (gdude2002) [inline bot/cogs/math.py]: No `setup(bot)` function

**Comment by martmists** [comment]
`__init__` is not actually required, but I was requested to add it for consistency

File: bot/cogs/math.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+LATEX_URL = "https://latex.codecogs.com/png.download?%5Cdpi%7B150%7D%20%5Cbg_white%20%5Chuge%20"
+
+
+class Math:
+    def __init__(self, bot):
+        self.bot = bot
+
+    @command()
+    async def latexify(self, ctx, expr: str):
```
Comment (gdude2002) [inline bot/cogs/math.py]: Could shove a splat before `expr` so that the expression can contain spaces?

**Comment by AnonGuy** [comment]
Not a must-have, but consider using a POST request to http://latex2png.com/ with the following form data, where `latex` is the parsed LaTeX expression:
image
A HTML colour of `808080` guarantees that the transparent image can be viewed on both themes, and you'll find the image URL here:
image
Again, not exactly necessary, but we'll get a much nicer transparent image:
image

**Comment by AnonGuy** [comment]
This is looking great so far! A couple issues with the discord width limit, but the images look good.

**Comment by martmists** [comment]
Discord width limit is something I cannot change, unless any of you have ideas

File: bot/cogs/math.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+            file = File(fp=BytesIO(bytes_img), filename="latex.png")
+
+            await ctx.send(file=file)
+
+    @command()
+    async def calc(self, ctx, *, expr: str):
+        fixed_expr = expr.replace('^', '**').strip('`')
+
+        if any(x in fixed_expr for x in ("**", "__")):
+            await ctx.send(
```
Comment (gdude2002) [inline bot/cogs/math.py]: I would prefer
```py
return await ctx.send(
    ...
)
```

**Comment by martmists** [comment]
There is no need to return though, and either none or all branches should return if we want a consistent style.

**Comment by gdude2002** [comment]
Well, you _have a return there_ on the next line, so..

**Comment by martmists** [comment]
Oh really? My bad then

**Comment by gdude2002** [comment]
Unfortunately, the merge conflicts are too complex for GitHub to handle - I trust I can leave that with you?

**Comment by martmists** [comment]
How do I fix them?

**Comment by gdude2002** [comment]
Huh, you don't have a mergetool?

You'll have to merge master into your branch and deal with the conflicts as usual.

**Comment by martmists** [comment]
I never had any issues merging so I wouldn't really know how ^^"

**Pull request merged**